Sources/SwiftKafka/KafkaProducer.swift

```
@@ -14,21 +14,246 @@
+            keyBytes = key.readBytes(length: key.readableBytes)
+        } else {
+            keyBytes = nil
+        }
+
+        self.messageIDCounter += 1
+
+        let responseCode = message.value.withUnsafeReadableBytes { valueBuffer in
+
+            // Pass message over to librdkafka where it will be queued and sent to the Kafka Cluster.
+            // Returns 0 on success, error code otherwise.
+            return rd_kafka_produce(
+                topicHandle,
+                message.partition.rawValue,
+                RD_KAFKA_MSG_F_COPY,
+                UnsafeMutableRawPointer(mutating: valueBuffer.baseAddress),
+                valueBuffer.count,
+                keyBytes,
+                keyBytes?.count ?? 0,
+                UnsafeMutableRawPointer(bitPattern: messageIDCounter)
+            )
+        }
+
+        guard responseCode == 0 else {
+            throw KafkaError(rawValue: responseCode)
+        }
+
+        return self.messageIDCounter
+    }
+
+    // Closure that is executed when a message has been acknowledged by Kafka
+    private let deliveryReportCallback: (
+        @convention(c) (OpaquePointer?, UnsafePointer<rd_kafka_message_t>?, UnsafeMutableRawPointer?) -> Void
+    ) = { _, messagePointer, producerPointer in
+        guard let producerPointer = producerPointer else {
+            fatalError("Could not resolve reference to KafkaProducer instance")
+        }
+        // Get a reference to the KafkaProducer without decrementing its reference count
+        let producer = Unmanaged<KafkaProducer>.fromOpaque(producerPointer).takeUnretainedValue()
+
+        guard let messagePointer = messagePointer else {
+            producer.logger.error("Could not resolved acknowledged message")
+            return
+        }
+
+        let messageID = messagePointer.pointee._private
+        guard let message = KafkaAcknowledgedMessage(messagePointer: messagePointer, id: UInt(bitPattern: messageID)) else {
```

I already commented on the `KafkaAcknowledgedMessage` that we should make the `init` throwing and not return `nil`. Furthermore, we should also surface this error in the `AsyncSequence` otherwise users will never know what went wrong. What about having a `Result` type as the element of the `AsyncSequence` stream and we also introduce a `KafkaAcknowledgedMessageError` that contains the id of the message the actual error?